### ISSUE TITLE
lua_function= doesn't work in [effect][filter]

### ISSUE THREAD

**Issue by gfgtdf** [wesnoth/wesnoth#3396]
`lua_function=` doesn't work in `[effect][filter]`. The reason is that lua_function tries to pass the unit by map location or underlying_id whihc will fail in that case becasue those are unit not yet copied to the map, (it uses  advancement.cpp::get_advanced_unit() which copies the unit.).

The workaround that i now use in an addon that i maintain addons is to define a custom effect that applies the lua filter, see:

```
function wesnoth.effects.wc2_fixed_lua_filter(u, cfg)
	local path = wc2_utils.split_to_array(cfg.lua_filter)
	local current = _G
	for v in tostring(cfg.lua_filter or ""):gmatch("[^%s.][^.]*") do
		current = current[v]
	end
	if current(u) then
		cfg.apply_to = cfg.apply_to2
		wesnoth.add_modification(u, "object", {
			T.effect (cfg),
		}, false)
	end
end

...
...

		[effect]
			apply_to=wc2_fixed_lua_filter
			apply_to2=max_experience
			lua_filter="wc2_utils.has_no_advances"

			set=75
		[/effect]
```

**Comment by CelticMinstrel** [comment]
So make `lua_function` pass in the actual unit object instead of just its location?

**Comment by gfgtdf** [comment]
lua_function does take a unit object but that is a lua unit object where a unit is internally referenced by underlyingid and assumed to be on the map, (unlike the type of lua object that is used by the effect implementation that has a real reference to the c++ unit object.)

**Comment by CelticMinstrel** [comment]
Why can't you just pass the same `lua_unit` that was passed to the effect in the first place?

**Comment by gfgtdf** [comment]
we could just change the implementation in `game_lua_kernel::run_filter` to be the same as `game_lua_kernel::apply_effect` (in terms us lua_unit construction).

(also since units use shared_ptr in master iirc, also since the refactor in unit construction we could also make it use a unit_ptr instead of just a unit&).

**Comment by gfgtdf** [comment]
i wonder whether this fix can go into 1.14 . I mean on the one hand it's clearly a bug, but on the other hand it also a behaviour change.

**Comment by CelticMinstrel** [comment]
Judging by past cases of similar bugs, I'd say it can probably go into 1.14. I'd recommend getting an opinion from (at minimum) shadowm and zookeeper, though.
